src/patterns.jl

```diff
@@ -0,0 +1,131 @@
+"""
+    AbstractPattern
+
+Abstract supertype of all sparsity pattern representations.
+
+## Type hierarchy
+```
+AbstractPattern
+├── AbstractVectorPattern: used in GradientTracer, ConnectivityTracer
+│   └── IndexSetVectorPattern
+└── AbstractHessianPattern: used in HessianTracer
+    └── IndexSetHessianPattern
+```
+"""
+abstract type AbstractPattern end
+
+#==========================#
+# Utilities on AbstractSet #
+#==========================#
+
+myempty(::Type{S}) where {S<:AbstractSet} = S()
+seed(::Type{S}, i::Integer) where {S<:AbstractSet} = S(i)
+
+""""
+    product(a::S{T}, b::S{T})::S{Tuple{T,T}}
+
+Inner product of set-like inputs `a` and `b`.
+"""
+product(a::AbstractSet{I}, b::AbstractSet{I}) where {I<:Integer} =
+    Set((i, j) for i in a, j in b)
+
+function union_product!(
+    hessian::SH, gradient_x::SG, gradient_y::SG
+) where {I<:Integer,SG<:AbstractSet{I},SH<:AbstractSet{Tuple{I,I}}}
+    hxy = product(gradient_x, gradient_y)
+    return union!(hessian, hxy)
+end
+
+#=======================#
+# AbstractVectorPattern #
+#=======================#
+
+# For use with ConnectivityTracer and GradientTracer.
+
+"""
+    AbstractVectorPattern <: AbstractPattern
+
+Abstract supertype of sparsity patterns representing a vector.
+For use with [`ConnectivityTracer`](@ref) and [`GradientTracer`](@ref).
+
+## Expected interface
+
+* `myempty(::Type{MyPattern})`: return a pattern representing a new number (usually an empty pattern)
+* `seed(::Type{MyPattern}, i::Integer)`: return an pattern that only contains the given index `i`
+* `inputs(p::MyPattern)`: return non-zero indices `i` for use with `ConnectivityTracer`
+* `gradient(p::MyPattern)`: return non-zero indices `i` for use with `GradientTracer`
```

Comment:
Why differentiate between these?

Reply:
Because the pattern represents something different when used in a `ConnectivityTracer` or `GradientTracer`. I could just call it `vector`, but I feel like that's also an odd name for `ConnectivityTracer`.

Suggestions are welcome, I fully agree that a single function would be best!

Reply:
Will be addressed by removing `ConnectivityTracer` in a follow-up PR.